xdsl/dialects/omp.py

```diff
@@ -583,6 +585,76 @@ def num_block_args(self) -> int:
         return len(self.private_vars) + len(self.reduction_vars)
 
 
+@irdl_op_definition
+class DeclareReductionOp(IRDLOperation):
+    """
+    Implementation of upstream omp.declare_reduction
+    See external [documentation](https://mlir.llvm.org/docs/Dialects/OpenMPDialect/ODS/#ompdeclare_reduction-ompdeclarereductionop).
+    """
+
+    name = "omp.declare_reduction"
+
+    sym_name = prop_def(SymbolNameConstraint())
+    var_type = prop_def(TypeAttribute, prop_name="type")
+
+    alloc_region = region_def()
+    init_region = region_def()
+    reduction_region = region_def()
+    atomic_reduction_region = region_def()
+    cleanup_region = region_def()
+
+    traits = traits_def(IsolatedFromAbove(), SymbolOpInterface())
+
+    assembly_format = """
+        $sym_name `:` $type attr-dict-with-keyword
+        ( `alloc` $alloc_region^ )?
+        `init` $init_region
+        `combiner` $reduction_region
+        ( `atomic` $atomic_reduction_region^ )?
+        ( `cleanup` $cleanup_region^ )?
+    """
+
+    def verify_(self) -> None:
+        if len(self.alloc_region.blocks) > 1:
+            raise VerifyException(
+                f"{self.name} should have at most 1 block in alloc_region"
+            )
+
+
+@irdl_op_definition
+class PrivateClauseOp(IRDLOperation):
+    """
+    Implementation of upstream omp.private
+    See external [documentation](https://mlir.llvm.org/docs/Dialects/OpenMPDialect/ODS/#ompprivate-ompprivateclauseop).
+    """
+
+    name = "omp.private"
+
+    sym_name = prop_def(SymbolNameConstraint())
+    var_type = prop_def(TypeAttribute, prop_name="type")
+    data_sharing_type = prop_def(DataSharingClauseAttr)
+
+    alloc_region = region_def()
+    copy_region = region_def()
+    dealloc_region = region_def()
+
+    traits = traits_def(IsolatedFromAbove())
+
+    assembly_format = """
+        $data_sharing_type $sym_name `:` $type
+        `alloc` $alloc_region
+        (`copy` $copy_region^)?
+        (`dealloc` $dealloc_region^)?
```

Comment:
This looks very different to the upstream docs, what's going on? (upstream has `init` instead of `alloc` and its optional)